src/lib/util/options.ml

```diff
@@ -28,38 +28,90 @@
 (*                                                                        *)
 (**************************************************************************)
 
-(* Formatter declarations, getters and setters *)
-let fmt_std = ref Format.std_formatter
-let fmt_err = ref Format.err_formatter
-let fmt_wrn = ref Format.err_formatter
-let fmt_dbg = ref Format.err_formatter
-let fmt_mdl = ref Format.std_formatter
-let fmt_usc = ref Format.std_formatter
-
-let set_std_fmt f =
-  fmt_std := f;
-  fmt_mdl := f;
-  fmt_usc := f
-
-let set_err_fmt f =
-  fmt_err := f;
-  fmt_wrn := f;
-  fmt_dbg := f
-
-let get_fmt_std () = !fmt_std
-let get_fmt_err () = !fmt_err
-let get_fmt_wrn () = !fmt_wrn
-let get_fmt_dbg () = !fmt_dbg
-let get_fmt_mdl () = !fmt_mdl
-let get_fmt_usc () = !fmt_usc
-
-let set_fmt_std f = fmt_std := f
-let set_fmt_err f = fmt_err := f
-let set_fmt_wrn f = fmt_wrn := f
-let set_fmt_dbg f = fmt_dbg := f
-let set_fmt_mdl f = fmt_mdl := f
-let set_fmt_usc f = fmt_usc := f
-
+module Output = struct
+  type t =
+    | Stdout
+    | Stderr
+    | Channel of out_channel * Format.formatter
+    | Buffer of Buffer.t * Format.formatter
+    | Invalid
+
+  let create_buffer () =
+    let buf = Buffer.create 10 in
+    let fmt = Format.formatter_of_buffer buf in
+    Buffer (buf, fmt)
+
+  let create_channel = function
+    | "stdout" -> Stdout
+    | "stderr" -> Stderr
+    | str ->
+      let cout = open_out str in
+      let fmt = Format.formatter_of_out_channel cout in
+      Channel (cout, fmt)
+
+  let std_output = ref Stdout
+  let err_output = ref Stderr
+  let wrn_output = ref Stderr
+  let dbg_output = ref Stderr
+  let mdl_output = ref Stdout
+  let usc_output = ref Stdout
+
+  let contents = function
+    | Stdout | Stderr | Channel _ | Invalid -> ""
```

Comment:
This is not a very good idea and can only cause confusion if that function ever ends up called on a non-buffer output. At the very least this should raise an exception, but a better API design would be:

 - Replace `Buffer of Buffer.t * Format.formatter` with `Formatter of Format.formatter`
 - Remove `create_buffer` and replace it with `let of_formatter fmt = Formatter fmt`
 - Remove `contents` entirely
 - Users of `create_buffer` (i.e. the js_worker) are responsible for calling `formatter_of_buffer` and `Buffer.contents` directly (and maybe there should be a `pp_print_flush` somewhere?)

Reply:
I only flush the formatter. According to the documentation:
```
formatter_of_buffer b returns a new formatter writing to buffer b. At the end of pretty-printing, the formatter must be flushed using [Format.pp_print_flush](https://v2.ocaml.org/api/Format.html#VALpp_print_flush) or [Format.pp_print_newline](https://v2.ocaml.org/api/Format.html#VALpp_print_newline), to print all the pending material into the buffer.
``` 
it should be fine :)

Reply:
It should happen in the finalizer `close_all` but I forgot to add it in the js_worker. I will do it right now. Using `set-option` calls systematically the function `Output.close`.

Reply:
ooh I thought `Buffer.contents` calls the flush procedure!